src/modules/ekf2/EKF/aid_sources/magnetometer/mag_control.cpp

```diff
@@ -185,74 +180,47 @@ void Ekf::controlMagFusion()
 
-				if (mag_sample.reset || checkHaglYawResetReq() || wmm_updated) {
+				if (mag_sample.reset || checkHaglYawResetReq() || (wmm_updated && no_ne_aiding_or_pre_takeoff)) {
 					ECL_INFO("reset to %s", AID_SRC_NAME);
 					resetMagStates(_mag_lpf.getState(), _control_status.flags.mag_hdg || _control_status.flags.mag_3D);
 					aid_src.time_last_fuse = _time_delayed_us;
 
 				} else {
-					if (!_mag_decl_cov_reset) {
-						// After any magnetic field covariance reset event the earth field state
-						// covariances need to be corrected to incorporate knowledge of the declination
-						// before fusing magnetometer data to prevent rapid rotation of the earth field
-						// states for the first few observations.
-						fuseDeclination(0.02f);
-						_mag_decl_cov_reset = true;
-						fuseMag(mag_sample.mag, R_MAG, H, aid_src);
+					// The normal sequence is to fuse the magnetometer data first before fusing
+					// declination angle at a higher uncertainty to allow some learning of
+					// declination angle over time.
+					const bool update_all_states = _control_status.flags.mag_3D || _control_status.flags.mag_hdg;
+					const bool update_tilt = _control_status.flags.mag_3D;
+					fuseMag(mag_sample.mag, R_MAG, H, aid_src, update_all_states, update_tilt);
+
+					// the innovation variance contribution from the state covariances is negative which means the covariance matrix is badly conditioned
+					if (update_all_states && update_tilt) {
+						_fault_status.flags.bad_mag_x = (aid_src.innovation_variance[0] < aid_src.observation_variance[0]);
+						_fault_status.flags.bad_mag_y = (aid_src.innovation_variance[1] < aid_src.observation_variance[1]);
+						_fault_status.flags.bad_mag_z = (aid_src.innovation_variance[2] < aid_src.observation_variance[2]);
+					}
 
-					} else {
-						// The normal sequence is to fuse the magnetometer data first before fusing
-						// declination angle at a higher uncertainty to allow some learning of
-						// declination angle over time.
-						const bool update_all_states = _control_status.flags.mag_3D || _control_status.flags.mag_hdg;
-						const bool update_tilt = _control_status.flags.mag_3D;
-						fuseMag(mag_sample.mag, R_MAG, H, aid_src, update_all_states, update_tilt);
-
-						// the innovation variance contribution from the state covariances is negative which means the covariance matrix is badly conditioned
-						if (update_all_states && update_tilt) {
-							_fault_status.flags.bad_mag_x = (aid_src.innovation_variance[0] < aid_src.observation_variance[0]);
-							_fault_status.flags.bad_mag_y = (aid_src.innovation_variance[1] < aid_src.observation_variance[1]);
-							_fault_status.flags.bad_mag_z = (aid_src.innovation_variance[2] < aid_src.observation_variance[2]);
-						}
-
-						if (_control_status.flags.mag_dec) {
-							fuseDeclination(0.5f);
-						}
+					if (_control_status.flags.mag_dec) {
+						fuseDeclination(0.5f);
```

Comment:
Why 0.5?

Reply:
Good question. I would say that "it's enough to prevent the declination from drifting away without constraining it too much". The declination fusion is just a "weak restoring force" to keep the heading aligned on the long run.
If 0.5 is the perfect value, I don't know, but it seems fine when looking at logs.